### PR TITLE
Added formula for Caddy HTTP server

### PR DESCRIPTION
**Pull request by dpoggi**
https://caddyserver.com/

For additional reference see https://github.com/Homebrew/homebrew-binary/pull/212

### CODE REVIEW

**Comment by mistydemeo** [comment]
Thanks very much for the submission!

Would you be able to clean up the audit style notes the bots flagged?

```
== /usr/local/Library/Formula/caddy.rb ==
C:  1:  9: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C:  2:  9: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C:  5: 12: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C:  6:  7: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C:  7:  8: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C:  8:  8: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 10: 14: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 12: 15: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 13:  9: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 13: 61: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 16: 15: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 17:  9: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 17: 63: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 20: 15: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 21:  9: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 21: 67: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 24: 15: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 25:  9: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 25: 63: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 28: 15: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 29:  9: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 29: 69: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 32: 15: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 33:  9: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 33: 79: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 39:  9: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 40:  9: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 40: 19: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 41:  9: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 41: 44: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 41: 54: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 44: 12: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 44: 18: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 44: 27: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 44: 33: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 45: 17: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.
C: 46: 17: %w-literals should be delimited by [ and ]

 * A `test do` test block should be added
 * Stable: SHA1 checksums are deprecated, please use SHA256
 * `require 'formula'` is now unnecessary
```

**Comment by dpoggi** [comment]
Should be good now. Took a few minutes to find a test that works for both v0.6.0 and HEAD (only HEAD has a -version switch to make things easy).

File: Library/Formula/caddy.rb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+  go_resource "github.com/russross/blackfriday" do
+    url "https://github.com/russross/blackfriday.git", :revision => "4bed88b4fd00fbb66b49b0f38ed3dd0b902ab515"
+  end
+
+  go_resource "github.com/shurcooL/sanitized_anchor_name" do
+    url "https://github.com/shurcooL/sanitized_anchor_name.git", :revision => "8e87604bec3c645a4eeaee97dfec9f25811ff20d"
+  end
+
+  if build.head?
```
Comment (jacknagel) [inline Library/Formula/caddy.rb]: `head do`, and then add the head url in this block instead as well.

File: Library/Formula/caddy.rb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    Language::Go.stage_deps resources, buildpath/"src"
+
+    system "go", "build", "-o", "caddy"
+    bin.install "caddy"
+    doc.install %w[README.md LICENSE.md]
+  end
+
+  test do
+    HOMEBREW_REPOSITORY.cd do
+      io = IO.popen("bin/caddy")
```
Comment (xu-cheng) [inline Library/Formula/caddy.rb]: `#{bin}/caddy`

File: Library/Formula/caddy.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    ENV["GOARCH"] = MacOS.prefer_64_bit? ? "amd64" : "386"
+    Language::Go.stage_deps resources, buildpath/"src"
+
+    system "go", "build", "-o", "caddy"
+    bin.install "caddy"
+    doc.install %w[README.md LICENSE.md]
+  end
+
+  test do
+    HOMEBREW_REPOSITORY.cd do
```
Comment (xu-cheng) [inline Library/Formula/caddy.rb]: Why change directory?

File: Library/Formula/caddy.rb [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      io = IO.popen("bin/caddy")
+      sleep 2
+      Process.kill("SIGINT", io.pid)
+      Process.wait(io.pid)
```
Comment (xu-cheng) [inline Library/Formula/caddy.rb]: `Process.kill` and `Process.wait` should be inside a `ensure` block.

File: Library/Formula/caddy.rb (hunk 3 above)
Comment (dpoggi) [inline Library/Formula/caddy.rb]: brew was coming up with the wrong cellar directory when HEAD was installed (it was going for caddy/0.6.0 still)

File: Library/Formula/caddy.rb (hunk 2 above)
Comment (dpoggi) [inline Library/Formula/caddy.rb]: It's bin/caddy instead of #{bin}/caddy because of the directory change, see the above line note.

**Comment by DomT4** [comment]
@BrewTestBot test this please

**Comment by DomT4** [comment]
This is green now, but if one of @Homebrew/owners could :+1: that'd be appreciated.

**Comment by cloudbring** [comment]
## Could someone (@Homebrew/owners) please approve this?
I'd love to use Caddy and just install it with `brew`

**Comment by tdsmith** [comment]
No gifs, please, but thanks for your enthusiasm.

**Comment by tdsmith** [comment]
@BrewTestBot please test this

**Comment by tdsmith** [comment]
LGTM once the bottles are ready.

**Comment by tdsmith** [comment]
@BrewTestBot test this please

File: Library/Formula/caddy.rb [hunk 5]
```diff
@@ -0,0 +1,70 @@
+require "language/go"
+
+class Caddy < Formula
+  homepage "https://caddyserver.com/"
```
Comment (bfontaine) [inline Library/Formula/caddy.rb]: Coud you add a short description using `desc "…"`? Thanks!

**Comment by tdsmith** [comment]
I got it. Thanks for the PR!

**Comment by cloudbring** [comment]
Awesome! Thanks! Can't wait to use Caddy and HTTP 2.0 for all my stuff now.

**Pull request closed without merging**